src/functions.rs

```diff
@@ -134,6 +144,91 @@ impl<'a> FmtVisitor<'a> {
+                                   unsafety: &ast::Unsafety,
+                                   abi: &abi::Abi)
+        -> String
+    {
+        let where_clause = &generics.where_clause;
+        let mut result = String::with_capacity(1024);
+
+        if let &ast::Unsafety::Unsafe = unsafety {
+            result.push_str("unsafe ");
+        }
+        if *abi != abi::Rust {
+            result.push_str("extern ");
+            result.push_str(&abi.to_string());
+            result.push(' ');
+        }
+
+        // fn foo
+        result.push_str("fn ");
+        result.push_str(&token::get_ident(ident));
+
+        // Generics.
+        let generics_indent = indent + result.len();
+        result.push_str(&self.rewrite_generics(generics,
+                                               generics_indent,
+                                               span_for_return(&fd.output)));
+
+        let ret_str = self.rewrite_return(&fd.output);
+
+        // Args.
+        let (one_line_budget, multi_line_budget, arg_indent) =
+            self.compute_budgets_for_args(&mut result, indent, ret_str.len(), false);
+
+        debug!("rewrite_fn: one_line_budget: {}, multi_line_budget: {}, arg_indent: {}",
+               one_line_budget, multi_line_budget, arg_indent);
+
+        result.push('(');
+        result.push_str(&self.rewrite_args(&fd.inputs,
+                                           explicit_self,
+                                           one_line_budget,
+                                           multi_line_budget,
+                                           arg_indent,
+                                           span_for_return(&fd.output)));
+        result.push(')');
+
+        // Return type.
+        if ret_str.len() > 0 {
+            // If we've already gone multi-line, or the return type would push
+            // over the max width, then put the return type on a new line.
+            if result.contains("\n") ||
+               result.len() + indent + ret_str.len() > MAX_WIDTH {
+                let indent = match FN_RETURN_INDENT {
+                    ReturnIndent::WithWhereClause => indent + 4,
+                    // TODO we might want to check that using the arg indent doesn't
+                    // blow our budget, and if it does, then fallback to the where
+                    // clause indent.
+                    _ => arg_indent,
+                };
+
+                result.push('\n');
+                result.push_str(&make_indent(indent));
+            } else {
+                result.push(' ');
+            }
+            result.push_str(&ret_str);
+        }
+
+        // Where clause.
+        let predicate_count = where_clause.predicates.len();
+        if predicate_count > 0 {
+            let comments = (0..predicate_count).map(|_| "".to_string()).collect();
+            result.push_str(&self.rewrite_where_clause(where_clause, indent, comments));
+        }
```

Comment:
I couldn't figure out how to do it, since it requires the next span and we do not have that in this case..